fe/fe-core/src/main/java/com/starrocks/load/InsertOverwriteJobRunner.java

```diff
@@ -0,0 +1,340 @@
+
+public class InsertOverwriteJobRunner {
+    private static final Logger LOG = LogManager.getLogger(InsertOverwriteJobRunner.class);
+
+    private InsertOverwriteJob job;
+
+    private long watershedTxnId = -1;
+    private InsertStmt insertStmt;
+    private StmtExecutor stmtExecutor;
+    private ConnectContext context;
+    private Database db;
+    private OlapTable targetTable;
+    private String postfix;
+
+    private long createPartitionElapse;
+    private long waitInsertIntoElapse;
+    private long insertElapse;
+
+    public InsertOverwriteJobRunner(InsertOverwriteJob job, ConnectContext context, StmtExecutor stmtExecutor,
+                                    Database db, OlapTable targetTable) {
+        this.job = job;
+        this.context = context;
+        this.stmtExecutor = stmtExecutor;
+        this.insertStmt = job.getInsertStmt();
+        this.db = db;
+        this.targetTable = targetTable;
+        this.postfix = "_" + job.getJobId();
+        this.createPartitionElapse = 0;
+        this.waitInsertIntoElapse = 0;
+        this.insertElapse = 0;
+    }
+
+    // for replay
+    public InsertOverwriteJobRunner(InsertOverwriteJob job) {
+        this.job = job;
+        this.db = GlobalStateMgr.getCurrentState().getDb(job.getTargetDbId());
+        this.targetTable = (OlapTable) db.getTable(job.getTargetTableId());
+        this.postfix = "_" + job.getJobId();
+        this.createPartitionElapse = 0;
+        this.waitInsertIntoElapse = 0;
+        this.insertElapse = 0;
+    }
+
+    public boolean isFinished() {
+        return job.isFinished();
+    }
+
+    // only called from log replay
+    // there is no concurrent problem here
+    public void cancel() {
+        if (isFinished()) {
+            LOG.warn("cancel failed. insert overwrite job:{} already finished. state:{}", job.getJobState());
+            return;
+        }
+        try {
+            transferTo(InsertOverwriteJobState.OVERWRITE_CANCELLED);
+        } catch (Exception e) {
+            LOG.warn("cancel insert overwrite job:{} failed", job.getJobId(), e);
+        }
+    }
+
+    public void run() throws Exception {
+        try {
+            handle();
+        } catch (Exception e) {
+            LOG.warn("insert overwrite job:{} handle exception", job.getJobId(), e);
+            if (job.getJobState() != InsertOverwriteJobState.OVERWRITE_FAILED
+                    && job.getJobState() != InsertOverwriteJobState.OVERWRITE_CANCELLED) {
+                transferTo(InsertOverwriteJobState.OVERWRITE_FAILED);
+            }
+            throw new RuntimeException("insert overwrite failed.", e);
+        }
+    }
+
+    public void handle() throws Exception {
+        switch (job.getJobState()) {
+            case OVERWRITE_PENDING:
+                prepare();
+                break;
+            case OVERWRITE_PREPARED:
+                doLoad();
+                break;
+            case OVERWRITE_FAILED:
+            case OVERWRITE_CANCELLED:
```

Comment:
maybe use OVERWRITE_FAILED instead of OVERWRITE_CANCELLED will be better?

Reply:
ok